node/src/node/isometrics.cpp

```diff
@@ -0,0 +1,528 @@
+#include <tuple>
+#include <cassert>
+
+#include "isometrics.hpp"
+
+#include <boost/range/adaptors.hpp>
+#include <boost/range/algorithm.hpp>
+#include <boost/range/numeric.hpp>
+
+// #define ISOMETRICS_DEBUG
+#undef ISOMETRICS_DEBUG
+
+#ifdef ISOMETRICS_DEBUG
+#define dbg(msg) std::cerr << msg << '\n'
+#define DBG_DUMP_UUIDS(os, logo, container) aux::dump_uuids(os, logo, container)
+#else
+#define dbg(msg)
+#define DBG_DUMP_UUIDS(os, logo, container)
+#endif
+
+namespace cocaine {
+namespace detail {
+namespace service {
+namespace node {
+
+namespace ph = std::placeholders;
+
+namespace conf {
+    constexpr auto PURGATORY_QUEUE_BOUND = 8 * 1024;
+    constexpr auto SHOW_ERRORS_LIMIT = 3;
+
+    // <name, is metric monotonically increasing per application (aggregate) level>
+    const std::vector<std::pair<std::string, bool>> counter_metrics_names =
+    {
+        // yet abstract cpu load measurement
+        { "cpu", false},
+
+        // memory usage (in bytes)
+        { "vms", false },
+        { "rss", false },
+
+        // running times
+        { "uptime",    true },
+        { "user_time", true },
+        { "sys_time",  true },
+
+        // disk io (in bytes)
+        { "ioread",  true },
+        { "iowrite", true },
+
+        // TODO: network stats
+
+    };
+}
+
+namespace aux {
+
+    template<typename Val>
+    auto
+    uuid_value(const Val& v) -> std::string;
+
+    template<>
+    auto
+    uuid_value(const std::string& v) -> std::string {
+        return v;
+    }
+
+    template<>
+    auto
+    uuid_value(const dynamic_t& arr) -> std::string {
+        return arr.as_string();
+    }
+
+    template<typename Container>
+    auto
+    dump_uuids(std::ostream& os, const std::string& logo, const Container& arr) -> void {
+        os << logo << " size " << arr.size() << '\n';
+        for(const auto& uuid : arr) {
+            os << "\tuuid: " << uuid_value(uuid) << '\n';
+        }
+    }
+}
+
+auto
+metrics_aggregate_proxy_t::operator+(const worker_metrics_t& worker_metrics) -> metrics_aggregate_proxy_t&
+{
+    using boost::adaptors::map_keys;
+    using counters_record_type = worker_metrics_t::counters_record_type;
+
+    dbg("worker_metrics_t::operator+() summing to proxy");
+
+    boost::for_each(worker_metrics.common_counters, [&](const counters_record_type &worker_record) {
+        const auto& name = worker_record.first;
+        const auto& metric = worker_record.second;
+
+        auto& self_record = this->common_counters[name];
+
+        self_record.values += metric.value->load();
+        self_record.deltas += metric.delta;
+    });
+
+    return *this;
+}
+
+auto
+operator+(worker_metrics_t& src, metrics_aggregate_proxy_t& proxy) -> metrics_aggregate_proxy_t&
+{
+    return proxy + src;
+}
+
+auto
+worker_metrics_t::operator=(metrics_aggregate_proxy_t&& proxy) -> worker_metrics_t& {
```

Comment:
Unobvious assignment operator usage with non-equal `this` and `other` types. Use method instead.